File: leetcode/java/0057. Insert Interval.java
package leetcode.java;

import java.util.ArrayList;
import java.util.List;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();

        for (int i=0; i<intervals.length; i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            if (newInterval[1] < start) {
                res.add(newInterval);
                for (int j=i; j<intervals.length; j++) {
                    res.add(intervals[j]);
                }
                return res.stream().toArray(int[][]::new);
            } else if (newInterval[0] > end) {
                res.add(intervals[i]);
            } else {
                newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
                newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
            }
        }
        res.add(newInterval);

        return res.stream().toArray(int[][]::new);
    }
}

File: leetcode/python/0057. Insert Interval.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[List[int]]
        :type newInterval: List[int]
        :rtype: List[List[int]]
        """

        res = []

        for idx in range(len(intervals)):
            start, end = intervals[idx]

            if start > newInterval[1]:
                res.append(newInterval)
                return res + intervals[idx:]
            elif end < newInterval[0]:
                res.append(intervals[idx])
            else:
                newInterval[0] = min(newInterval[0], start)
                newInterval[1] = max(newInterval[1], end)
        res.append(newInterval)
        return res